### ISSUE TITLE
java.lang.ClassCastException: class java.util.concurrent.TimeoutException cannot be cast to class xsbti.FullReload

### ISSUE THREAD

**Issue by tolikbotov** [sbt/sbt#6592]
## steps

Project build on heroku Stack-20. It's ubuntu 20.04. Java 11.0
Part from logs:
`[info] welcome to sbt 1.5.4 (Oracle Corporation Java 11.0.11)`

Plugin: `addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.8.1")`

Project is multimodule:
```
lazy val core = project.dependsOn(common, migration % "test->test")
  .enablePlugins(JavaAppPackaging)

lazy val migration = project.dependsOn(common % "test->test;compile->compile")
  .enablePlugins(JavaAppPackaging)

lazy val common = project
```

## problem

When compile is over there is an error:
```
-----> Running: sbt compile stage
...
[warn] one warning found
       [info] done compiling
       [success] Total time: 79 s (01:19), completed Jul 12, 2021, 9:12:02 AM
sbt appears to be exiting abnormally.
  The log file for this session is at /tmp/scala_buildpack_build_dir/target/global-logging/sbt-global-log1336679964453179936.log
java.lang.ClassCastException: class java.util.concurrent.TimeoutException cannot be cast to class xsbti.FullReload (java.util.concurrent.TimeoutException is in module java.base of loader 'bootstrap'; xsbti.FullReload is in unnamed module of loader 'app')
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:59)
	at sbt.xMain.run(Main.scala:46)
	at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
	at xsbt.boot.Launch$.run(Launch.scala:149)
	at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
	at xsbt.boot.Launch$.launch(Launch.scala:159)
	at xsbt.boot.Launch$.apply(Launch.scala:44)
	at xsbt.boot.Launch$.apply(Launch.scala:21)
	at xsbt.boot.Boot$.runImpl(Boot.scala:78)
	at xsbt.boot.Boot$.run(Boot.scala:73)
	at xsbt.boot.Boot$.main(Boot.scala:21)
	at xsbt.boot.Boot.main(Boot.scala)
[error] [launcher] error during sbt launcher: java.lang.ClassCastException: class java.util.concurrent.TimeoutException cannot be cast to class xsbti.FullReload (java.util.concurrent.TimeoutException is in module java.base of loader 'bootstrap'; xsbti.FullReload is in unnamed module of loader 'app')
```
## expectation

I think, after compilation sbt should generate documentation and pack all with stage command.

## notes

On previous stack-18 it compiles normally:
```
Description:	Ubuntu 18.04.5 LTS
Release:	18.04
```

**Comment by DougC** [comment]
I'm seeing the same problem with sbt 1.5.4 and 1.5.5 and jdk 11. Build works fine with jdk 8.

**Comment by eed3si9n** [comment]
@tolikbotov Thanks for the report!

**Comment by Malax** [comment]
I'm the maintainer of the Scala buildpack at Heroku, I don't think this is an issue with our buildpack, but if there is anything I can do to aid debugging, and/or fixing the issue let me know, happy to help! :)

**Comment by DougC** [comment]
@Malax it may have something to do with the build environment. When I run `sbt compile stage` (which I believe is the command the buildpack runs) with jdk 11 then sbt exits without a problem. Does the heroku build environment provide any extra plugins or anything?

**Comment by Malax** [comment]
@DougC There is indeed a plugin that we add when you use the buildpack: https://github.com/heroku/heroku-buildpack-scala/blob/f213cdcdd3f62ba211a2aac5507d1a097c7b2c8c/opt/HerokuBuildpackPlugin_sbt1.scala

Other that that, the buildpack heavily relies on [sbt-extras](https://github.com/dwijnand/sbt-extras) to do the heavy lifting. Do you see your issue on Heroku as well?

**Comment by DougC** [comment]
@Malax Yes. I only see the issue when building on Heroku with jdk 11. Build is fine with jdk 8.

Just tried building on my local machine with sbt-extras and it seems fine. Strange one.

**Comment by eed3si9n** [comment]
I am guessing this is a collaborative bug between the environment and sbt.
Something in the environment is making a background thread to throw `TimeoutException`, and instead of ignoring it there's a casting bug that I made back in January that's casting to `FullReload`:

https://github.com/sbt/sbt/blob/d4162cce04615e4267b0582d8a0f01170f3ba5af/main/src/main/java/sbt/internal/XMainConfiguration.java#L59

Not sure why it happens only on JDK 11.

**Comment by tolikbotov** [comment]
On my local machine 
```
         anatoliibotov@MacBook-Pro-Anatolij ~ % java -version
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
```

`sbt compile stage` works as expected. Only fails on Heroku.

Heroku:
```
~ $ java -version
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment (build 11.0.11+9)
OpenJDK 64-Bit Server VM (build 11.0.11+9, mixed mode)
```

Can JVM work differently? It seems no.

**Comment by rpiaggio** [comment]
We are having the same issue. We reverted to Stack-18 for the moment. We didn't try with Java 8.

**Comment by Malax** [comment]
@rpiaggio do you have a standalone reproduce case by chance? I spoke to @tolikbotov but wasn't able to reproduce on my end when they reported the issue to me initially. I'm sure @eed3si9n is right here, it looks like a collaborative bug and I'm happy to do my part to get it fixed. :)

**Comment by rpiaggio** [comment]
Hi @Malax ! Thank you. 

I was able to reproduce the issue by creating a new Heroku app, cloning https://github.com/gemini-hlsw/explore and then attempting to push-deploy it (master branch).

**Comment by tolikbotov** [comment]
@Malax is there any news?

**Comment by Liodene** [comment]
@Malax We're having the same issue using Heroku.
We were originally on sbt 1.3.10 and had no issues.

We've tried to upgrade to 1.5.5 and got the java.lang.ClassCastException: class java.util.concurrent.TimeoutException cannot be cast to class xsbti.FullReload,
Then we've downgraded to 1.4.9 and got the original TimeoutException only.
`sbt appears to be exiting abnormally.
  The log file for this session is at /tmp/scala_buildpack_build_dir/target/global-logging/sbt-global-log14917995661549164484.log
java.util.concurrent.TimeoutException
	at sbt.internal.TaskProgress.close(TaskProgress.scala:75)
	at sbt.MainLoop$.next(MainLoop.scala:200)
	at sbt.MainLoop$.run(MainLoop.scala:144)
	at sbt.MainLoop$.$anonfun$runWithNewLog$1(MainLoop.scala:119)
	at sbt.io.Using.apply(Using.scala:27)`

Now, we're back to 1.3.10 and it's working like a charm.
We tried to build our project on multiple envs with JDK 11 (local macOS, local windows 10, CircleCI) and it works everywhere but on Heroku.

Apparently, there's an issue to close an executor properly/in time on Heroku's env for some reason. The TaskProgress' method was rewritten totally between 1.3.10 and 1.4.9 and doesn't seem to rely on Executor (on the surface) so I guess that explains why it works on 1.3.10.

**Comment by aspantel** [comment]
@Malax any chance this will be fixed for Heroku? Thanks

**Comment by rubin55** [comment]
This does not *only* occur on Heroku. I'm on Windows 10 ARM64, using Microsoft's Java (for ARM64) version 16.0.1. When using sbt 1.5.5, I'm getting the same class casting error (i.e., casting to xsbti.FullReload, however, in my case it seems to be trying to cast NoClassDefFoundError to xsbti.FullReload?). The output (this is after succesfully executing `sbt new scala/scala3.g8`):

```
C:\Users\rubin\Desktop\scalaexample>sbt
Sep 01, 2021 8:59:40 PM org.jline.utils.Log logr
WARNING: Unable to create a system terminal, creating a dumb terminal (enable debug logging for more information)
←[0Jjava.lang.ClassCastException: class java.lang.NoClassDefFoundError cannot be cast to class xsbti.FullReload (java.lang.NoClassDefFoundError is in module java.base of loader 'bootstrap'; xsbti.FullReload is in unnamed module of loader 'app')
        at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:59)
        at sbt.xMain.run(Main.scala:46)
        at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
        at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
        at xsbt.boot.Launch$.run(Launch.scala:149)
        at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
        at xsbt.boot.Launch$.launch(Launch.scala:159)
        at xsbt.boot.Launch$.apply(Launch.scala:44)
        at xsbt.boot.Launch$.apply(Launch.scala:21)
        at xsbt.boot.Boot$.runImpl(Boot.scala:78)
        at xsbt.boot.Boot$.run(Boot.scala:73)
        at xsbt.boot.Boot$.main(Boot.scala:21)
        at xsbt.boot.Boot.main(Boot.scala)
[error] [launcher] error during sbt launcher: java.lang.ClassCastException: class java.lang.NoClassDefFoundError cannot be cast to class xsbti.FullReload (java.lang.NoClassDefFoundError is in module java.base of loader 'bootstrap'; xsbti.FullReload is in unnamed module of loader 'app')
```

I'm not 100% sure this is the same issue, though it might be related. If not, let me know and I'll make a new issue for it.

**Comment by Liodene** [comment]
The casting bug was made by @eed3si9n according to him above. The root causes are different though (TimeoutException vs NoClassDefFoundError),

**Comment by eed3si9n** [comment]
Yea. I could probably fix the casting problem, but it would just throw the underlying errors like TimeoutException or NoClassDefFoundError.

**Comment by eed3si9n** [comment]
https://github.com/sbt/sbt/pull/6649

**Comment by rubin55** [comment]
> Yea. I could probably fix the casting problem, but it would just throw the underlying errors like TimeoutException or NoClassDefFoundError.

Would that at least obtain the underlying reason for the NoClassDefFound in my case?

**Comment by hvesalai** [comment]
Got this same

```[error] [launcher] error during sbt launcher: java.lang.ClassCastException: class java.util.concurrent.TimeoutException cannot be cast to class xsbti.FullReload (java.util.concurrent.TimeoutException is in module java.base of loader 'bootstrap'; xsbti.FullReload is in unnamed module of loader 'app') ```

**Comment by eed3si9n** [comment]
Could you try sbt 1.6.0-M1?

**Comment by hvesalai** [comment]
@eed3si9n 1.6.0-M1 works correctly

```
[error] ## Exception when compiling 38 sources to....
...
[error] [launcher] error during sbt launcher: java.lang.OutOfMemoryError: Direct buffer memory
```

(OOME and TimeoutException worked the same before and OOME is easier to repro)

**Comment by eed3si9n** [comment]
Here's another PR https://github.com/sbt/sbt/pull/6729 based on the hint provided by @rtyley.

**Comment by Malax** [comment]
@eed3si9n: Thanks for looking into this from the sbt side! I appreciate the work you put into this, especially since it turned out to be triggered by the specific Heroku build environment and does not seem to be a bug of `sbt` itself.

@DougC, @tolikbotov, @rpiaggio, @Liodene, @yoxel: After some long debugging sessions, I eventually found the root cause for this issue on Heroku. Our OpenJDK 11 builds for the `heroku-20` stack triggered this behaviour in `sbt`. I found that the default `gcc` for this stack we used to build the JDK with, overzealously optimised some part of the JDK, triggering this behaviour.

I just released updated binaries that should no longer cause this problem - they will be used automatically during your next build with OpenJDK 11 on `heroku-20` and your sbt builds should finish without running into timeouts. I'm sorry that it took so long to get this resolved. :-(
